### PR TITLE
Return evaluated properties in query results

### PR DESCRIPTION
**Pull request by ryanhamley**
## Launch Checklist

 - [x] briefly describe the changes in this PR
    - closes https://github.com/mapbox/mapbox-gl-js/issues/8373
    - serialize the layer once and cache it to avoid serializing on every feature or tile
    - evaluate properties and return their values to in the query
 - [ ] write tests for all new functionality
 - [ ] document any changes to public APIs
 - [x] post benchmark scores
 - [x] manually test the debug page
 - [ ] tagged `@mapbox/gl-native` if this PR includes shader changes or needs a native port
 - [ ] apply changelog label ('bug', 'feature', 'docs', etc) or use the label 'skip changelog'
 - [ ] add an entry inside this element for inclusion in the `mapbox-gl-js` changelog: `<changelog></changelog>`

### CODE REVIEW

**Comment by ryanhamley** [comment]
For some reason, when no geometry is supplied to the query, this update is consistently faster than `master`, but when the geometry is supplied, it's consistently slower. I'm investigating that.
Screen Shot 2020-01-27 at 2 58 07 PM

Screen Shot 2020-01-27 at 3 03 22 PM

**Comment by chloekraw** [comment]
cc @pozdnyakov @mapbox/gl-native

**Comment by ryanhamley** [comment]
After exploring an alternate approach in #9282 and testing out a branch that broke out layer serialization on its own (#9289), I discovered that the bulk of the performance hit in the `QueryPoint` benchmark is from serializing the layers in `queryRenderedFeatures`. If we serialize each layer as it's added and cache the results, we can get the benchmark almost to where `master` is as shown below. A test that removed the call to `listImages` in `queryRenderedFeatures` eliminated the remaining performance hit so the next step is to cache the list of availableImages.

Screen Shot 2020-02-11 at 6 42 19 PM

I ran the full benchmark suite against #9289 and didn't find any significant performance degradation from doing so while `QueryBox` showed a significant improvement.

**Comment by ryanhamley** [comment]
Now that I've cached `serializedLayers` and `availableImages`, the benchmarks are all back inline with `master` and `queryBox` shows a marked improvement, which was the original goal of the ticket.

Screen Shot 2020-02-12 at 12 32 29 PM
Screen Shot 2020-02-12 at 12 38 45 PM
Screen Shot 2020-02-12 at 12 32 37 PM
Screen Shot 2020-02-12 at 12 32 48 PM
Screen Shot 2020-02-12 at 12 32 59 PM

File: src/data/feature_index.js [hunk 1]
```diff
@@ -269,6 +295,16 @@ register(
...
 
+function evaluateProperties(serializedProperties, styleLayerProperties, evaluationParameters) {
+    const evaluatedProperties = {};
+    for (const property in serializedProperties) {
+        const prop = styleLayerProperties instanceof PossiblyEvaluated ? styleLayerProperties.get(property) : null;
+        evaluatedProperties[property] = prop && prop.evaluate ? prop.evaluate(evaluationParameters.feature, evaluationParameters.featureState, evaluationParameters.availableImages) : prop;
+    }
+
+    return evaluatedProperties;
+}
```
Comment (mourner) [inline src/data/feature_index.js]: Nit: we have a `mapObject` function in `util`, so you could simplify this code to just the callback function for mapping.

File: src/data/feature_index.js [hunk 2]
```diff
@@ -194,16 +198,34 @@ class FeatureIndex {
...
+                featureState = sourceFeatureState.getState(styleLayer.sourceLayer || '_geojsonTileLayer', id);
+            }
+
+            const serializedLayer = serializedLayers[layerID];
+
+            const evaluationParameters = {
+                feature,
+                featureState,
+                availableImages
+            };
```
Comment (mourner) [inline src/data/feature_index.js]: Nit: probably no need to wrap those in an object if you immediately unwrap them in the lines below — separate arguments are fine.

File: src/style/style.js [hunk 3]
```diff
@@ -408,11 +415,11 @@ class Style extends Evented {
             this.sourceCaches[sourceId].used = false;
         }
 
-        const availableImages = this.imageManager.listImages();
+        // const availableImages = this.imageManager.listImages();
```
Comment (mourner) [inline src/style/style.js]: Nit: remove leftover

File: src/style-spec/expression/types/resolved_image.js [hunk 4]
```diff
@@ -18,8 +18,9 @@ export default class ResolvedImage {
         return this.name;
     }
 
-    static fromString(name: string): ResolvedImage {
-        return new ResolvedImage({name, available: false});
+    static fromString(name: string, availableImages?: ?Array<string>): ResolvedImage {
+        const available = availableImages ? availableImages.indexOf(name) > -1 : false;
+        return new ResolvedImage({name, available});
```
Comment (mourner) [inline src/style-spec/expression/types/resolved_image.js]: Is this change (including `coercion` expression) related to the PR, or an independent one?

Also noting `availableImages.indexOf` can have performance implications — how often is this method called and in what cases?

**Comment by mourner** [comment]
Also just realized this would benefit from some query tests that cover the new functionality.

**Comment by ryanhamley** [comment]
The benchmarks look good still. `QueryBox` shows a major performance improvement and everything else shows little to no change.

You can see a manual test I performed here where `"line-width": ["-", 13, 4]` is output as `"line-width": 9` 
Screen Shot 2020-02-27 at 7 13 38 PM

Screen Shot 2020-02-27 at 7 12 55 PM

@mourner do you have any thoughts on the best way to test this? our query tests aren't really set up to evaluate expressions and the unit tests for `style.js` use a mock `queryRenderedFeatures`.

**Comment by mourner** [comment]
@ryanhamley awesome! Please edit the PR to add the changelog item + label.

**Pull request merged**